### ISSUE TITLE
ENH: GeoDataFrame.cx should use the spatial index 

### ISSUE THREAD

**Issue by felix-ht** [geopandas/geopandas#2671]
#### Is your feature request related to a problem?

If i have a large dataset `GeoDataFrame.cx` is very slow in comparison to `sindex.contains`.

#### Describe the solution you'd like

`GeoDataFrame.cx` should use the sindex if it exists

#### API breaking implications

No api changes

#### Describe alternatives you've considered

` -`

#### Additional context

```python
import timeit
import geopandas as gpd
import random

from shapely.geometry import box as shapely_box


def make_geodataframe_with_random_points(
    count: int = 10, crs: str = "EPSG:4326", within: tuple = (0, 10, 0, 10)
) -> gpd.GeoDataFrame:
    random_points = [
        (random.uniform(within[0], within[1]), random.uniform(within[2], within[3]))
        for _ in range(count)
    ]
    gdf = gpd.GeoDataFrame(
        random_points,
        columns=["longitude", "latitude"],
        geometry=gpd.points_from_xy(*zip(*random_points)),
        crs=crs,
    )
    return gdf


points_df = make_geodataframe_with_random_points(count=100000, within=(0, 100, 0, 100))
points_df.sindex

duration_cx = timeit.timeit(lambda: points_df.cx[5:6, 5:6], number=100)

duration_sindex = timeit.timeit(
    lambda: points_df.sindex.query(shapely_box(5, 6, 5, 6), predicate="contains"),
    number=10,
)
print(f"Speedup: {duration_cx / duration_sindex}")

```

`>>> Speedup: 16634.362953391726`

**Comment by martinfleis** [comment]
Thanks for the suggestion! I remember doing some comparisons myself some time ago on polygons, where the difference was negligible. Points are clearly a different situation, especially with `contains` predicate. However, when the point is on the boundary of the box, it will return False, while `cx` returns True.

**Comment by jorisvandenbossche** [comment]
Indeed, we had an implementation and some benchmarks in https://github.com/geopandas/geopandas/pull/1949, which in the end was not merged. Of course, that was also only tested with one case, and it might depend a lot on the actual data (and the subset you are selecting).

I think your example is a case of selecting only a small subset (around 10 out of 100000), which was in the other PR also identified as the case where it actually _can_ be beneficial.

Also ensure you have either pygeos or shapely 2.0 installed as that makes a big difference. The timings in the other PR are assuming that is the case.

If I run your example, I still see a speedup, but only of around 100x instead of 2000x

**Comment by felix-ht** [comment]
Thanks for the great job you do on this project!

@jorisvandenbossche  If i install pygeos the speedup goes down to 66 times! 66x should still be worth the effort tho.

**Comment by felix-ht** [comment]
Here's a plot how the performance of the query changes over dataset size. Especially for large datasets the difference is dramatic - as sindex seems to have constant 

![Figure_1](https://user-images.githubusercontent.com/24698503/207017020-f49793b2-30d4-4fc2-a566-a7b8df90b407.png)

![Figure_2](https://user-images.githubusercontent.com/24698503/207017016-3e6e9ce7-1cd8-4826-aa23-54f1d0189b70.png)

![Figure_3](https://user-images.githubusercontent.com/24698503/207017012-19f771de-f2c1-47eb-9f2b-61d4d5429695.png)

**Comment by jorisvandenbossche** [comment]
@felix-ht do you have the code to replicate those benchmarks? (from the plots in your last comment)
